### PR TITLE
[tests] Test attention backends

### PR DESCRIPTION
**Pull request by sayakpaul**
# What does this PR do?

Adds a lightweight test suite for popular attention backends. By default this won't be run on our CI.

### CODE REVIEW

File: tests/others/test_attention_backends.py [hunk 1]
```diff
@@ -0,0 +1,264 @@
...
+
+pytestmark = pytest.mark.skipif(
+    os.getenv("RUN_ATTENTION_BACKEND_TESTS", "false") == "false", reason="Feature not mature enough."
+)
+from diffusers import FluxPipeline  # noqa: E402
+from diffusers.utils import is_torch_version  # noqa: E402
+
+
+FORWARD_CASES = [
+    ("flash_hub", None),
```
Comment (sayakpaul) [inline tests/others/test_attention_backends.py:34]: Will add this once https://github.com/huggingface/diffusers/pull/12387 is merged.

File: tests/others/test_attention_backends.py [hunk 2]
```diff
@@ -0,0 +1,264 @@
...
+                0.2441,
+                0.2695,
+            ],
+            dtype=torch.bfloat16,
+        ),
+    ),
+]
+
+COMPILE_CASES = [
+    ("flash_hub", None, True),
```
Comment (sayakpaul) [inline tests/others/test_attention_backends.py:50]: Will add add the test slices after https://github.com/huggingface/diffusers/pull/12387 is merged.

File: tests/others/test_attention_backends.py [hunk 3]
```diff
@@ -0,0 +1,264 @@
...
+    return torch.device("cuda:0")
+
+
+@pytest.fixture(scope="session")
+def pipe(device):
+    torch.set_grad_enabled(False)
+    model_id = "black-forest-labs/FLUX.1-dev"
+    pipe = FluxPipeline.from_pretrained(model_id, torch_dtype=torch.bfloat16).to(device)
+    pipe.set_progress_bar_config(disable=True)
+    pipe.transformer.eval()
```
Comment (sayakpaul) [inline tests/others/test_attention_backends.py]:
```suggestion
    repo_id = "black-forest-labs/FLUX.1-dev"
    pipe = FluxPipeline.from_pretrained(repo_id, torch_dtype=torch.bfloat16).to(device)
    pipe.set_progress_bar_config(disable=True)
```

**Comment by lauri9** [comment]
Hi! I'm implementing a new attention backend and in preparation for that I tried the unit tests from this PR. I was working in an environment with a different torch version from nightly `2.10.0.dev20250924+cu128` indicated in the unit test file. In my environment the native backend produces numerically divergent results from expected, as seen in the following pytest output snippet:
```
_____________________________________________________________________________________________________________________ test_forward_with_compile[native] ______________________________________________________________________________________________________________________
output = FluxPipelineOutput(images=tensor([[[[0.0391, 0.0391, 0.0410,  ..., 0.2090, 0.2090, 0.2070],
          [0.0391, 0.0586,...8],
          [0.0879, 0.0801, 0.0801,  ..., 0.2930, 0.2891, 0.3184]]]],
       device='cuda:0', dtype=torch.bfloat16))
expected_slice = tensor([0.0410, 0.0410, 0.0449, 0.0508, 0.0508, 0.0605, 0.0605, 0.0605, 0.2344,
        0.2461, 0.2578, 0.2773, 0.2871, 0.2832, 0.2773, 0.3066],
       dtype=torch.bfloat16)

    def _check_if_slices_match(output, expected_slice):
        img = output.images.detach().cpu()
        generated_slice = img.flatten()
        generated_slice = torch.cat([generated_slice[:8], generated_slice[-8:]])
>       assert torch.allclose(generated_slice, expected_slice, atol=1e-4)
E       assert False
E        +  where False = <built-in method allclose of type object at 0x7f4b31c218c0>(tensor([0.0391, 0.0391, 0.0410, 0.0488, 0.0449, 0.0566, 0.0586, 0.0566, 0.2422,\n        0.2539, 0.2656, 0.2871, 0.2969, 0.2930, 0.2891, 0.3184],\n       dtype=torch.bfloat16), tensor(
[0.0410, 0.0410, 0.0449, 0.0508, 0.0508, 0.0605, 0.0605, 0.0605, 0.2344,\n        0.2461, 0.2578, 0.2773, 0.2871, 0.2832, 0.2773, 0.3066],\n       dtype=torch.bfloat16), atol=0.0001)
E        +    where <built-in method allclose of type object at 0x7f4b31c218c0> = torch.allclose

tests/others/test_attention_backends.py:103: AssertionError
```
There were differences also in the eager mode tests.

Is it expected that the values diverge between versions? Could there be a better way to test than comparing numerical accuracy if the values are expected to vary between versions?

**Comment by sayakpaul** [comment]
The hardware might matter too. I ran the tests on an H100, actually. The CUDA version could matter, too. 

I will try to swap out exact assertion with cosine similarity-based checks which are a little more reliable and robust.

File: tests/others/test_attention_backends.py [hunk 4]
```diff
@@ -0,0 +1,144 @@
...
+
+
+@pytest.mark.parametrize(
+    "backend_name,expected_slice,error_on_recompile",
+    COMPILE_CASES,
+    ids=[c[0] for c in COMPILE_CASES],
+)
+def test_forward_with_compile(pipe, backend_name, expected_slice, error_on_recompile):
+    if "native" in backend_name and error_on_recompile and not is_torch_version(">=", "2.9.0"):
+        pytest.xfail(f"Test with {backend_name=} is compatible with a higher version of torch.")
```
Comment (dg845) [inline tests/others/test_attention_backends.py:128]:
```suggestion
        pytest.xfail(f"Test with {backend_name} is compatible with a higher version of torch.")
```
nit: typo?

**Review by dg845 — APPROVED** [review]
LGTM :)
